cpp/src/text/replace.cu

```diff
@@ -31,78 +31,161 @@ namespace nvtext {
+      d_offsets[idx] = nbytes;
+  }
+};
+
+using strings_iterator = cudf::column_device_view::const_iterator<cudf::string_view>;
+
+/**
+ * @brief Functor to replace tokens in each string.
+ *
+ * This tokenizes a string using the given d_delimiter and replaces any tokens that match
+ * a string in d_targets_begin/end with those from the d_replacements column.
+ * Strings with no matching tokens are left unchanged.
+ *
+ * This should be called first to compute the size of each output string and then a second
+ * time to fill in the allocated output buffer for each string.
+ */
+struct replace_tokens_fn : base_token_replacer_fn {
+  strings_iterator d_targets_begin;  ///< strings to search for
+  strings_iterator d_targets_end;
+  cudf::column_device_view const d_replacements;  ///< replacement strings
+
+  replace_tokens_fn(cudf::column_device_view const& d_strings,
+                    cudf::string_view const& d_delimiter,
+                    strings_iterator d_targets_begin,
+                    strings_iterator d_targets_end,
+                    cudf::column_device_view const& d_replacements)
+    : base_token_replacer_fn{d_strings, d_delimiter},
+      d_targets_begin{d_targets_begin},
+      d_targets_end{d_targets_end},
+      d_replacements{d_replacements}
+  {
+  }
+
+  /**
+   * @brief Return replacement string for the given token.
+   *
+   * @param token Token candidate to be replaced.
+   * @return result pair specifies replacement condition and new string
+   */
+  __device__ replace_result token_replacement(cudf::string_view const& token)
+  {
+    // check if the token matches any of the targets
+    auto const found_itr = thrust::find(thrust::seq, d_targets_begin, d_targets_end, token);
+    if (found_itr != d_targets_end) {  // match found
+      // retrieve the corresponding replacement string or
+      // if only one repl string, use that one for all targets
+      auto const d_repl = [&] {
+        auto const repl_idx = thrust::distance(d_targets_begin, found_itr);
+        return d_replacements.size() == 1 ? d_replacements.element<cudf::string_view>(0)
+                                          : d_replacements.element<cudf::string_view>(repl_idx);
+      }();
+      return replace_result{true, d_repl};
+    }
+    // otherwise, do not replace this token
+    return replace_result{false, cudf::string_view()};
+  }
+
+  __device__ void operator()(cudf::size_type idx)
+  {
+    process_string(
+      idx, [this] __device__(cudf::string_view const& token) { return token_replacement(token); });
+  }
+};
+
+/**
+ * @brief Functor to filter tokens in each string.
+ *
+ * This tokenizes a string using the given d_delimiter and replaces any tokens
+ * that are shorter than min_token_length with a replacement string.
+ *
+ * This should be called first to compute the size of each output string and then
+ * a second time to fill in the allocated output buffer for each string.
+ */
+struct filter_tokens_fn : base_token_replacer_fn {
```

Comment:
Since this class specifically filters short strings, I feel like the class name should reflect this.